### PR TITLE
Autosave: Improve E2E test reliability by consuming synchronous data and bailing on save failure

### PR DESCRIPTION
**Pull request by mcsf**
## Description

Recently-merged #17501 and #17503 greatly expanded the reach of autosaving and its E2E test coverage. These tests have been remarkably unreliable, especially on Travis, something which this PR sought to fix, and does fix in two main ways.

### Terminology preamble

A **local** autosave is an action — or its resulting object — performed on the client, the object then residing in `window.sessionStorage`. A **remote** autosave is a special kind of WordPress post revision saved on the WordPress site itself. Local autosaves are cheaper but more volatile, and can be performed more frequently, whereas remote autosaves are persisted across clients for a given WordPress user. Currently, per #17500, remote autosaves should take precedence over local ones._

### Fixes

1. How `LocalAutosaveMonitor` waits for remote autosave data.

* Until now, although we were requesting the entire remote autosave entity inside `LocalAutosaveMonitor`'s `useAutosaveNotice`, we only cared about the existence or absence of a remote autosave for the current post. `useAutosaveNotice` was requesting the entire remote entity because there were plans in #17501 to inspect and compare local and remote autosave objects, a requirement which was later dropped. This data needs to be resolved through the `core` data store.

* In contrast, the editor setting `autosave` is immediately available upon editor initialisation. It only contains an `editLink` property that links to a remote autosave's revision page, but most importantly it's an indication of whether there is a remote autosave available for the current post.

* Later on, _if_ we choose to reconcile remote and local autosaves with more sophistication, this could possibly be done based on recency of autosave — in which case we could might still get away with editor settings if the server provides this information (e.g. timestamp) inside the `autosave` object.

* This change should hopefully provide a more reliable page-loading experience for autosave's E2E tests.

2. How the `shouldn't conflict with server-side autosave` test bails under unexpected circumstances.

* Despite the changes described in 1., a particular test would still frequently fail. Debugging led to the conclusion that the test's forced remote autosaves might be failing for unknown reasons.

* This remote autosave failure is independent from the test's scope, which is to make sure that, concurrent remote and local autosaves for the current post, the editor will only show a notice to recover from the remote one.

* Thus, the temporary solution here is to let the test bail if, upon editor reload, it looks like there is no remote autosave available.

## How has this been tested?

E2E tests have been run extensively, and Travis should reliably green-light this branch.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by gziolo** [comment]
Travis isn't flattered with the proposed changes:
https://travis-ci.com/WordPress/gutenberg/jobs/240704049#L1010

File: packages/e2e-tests/specs/autosave.test.js [hunk 1]
```diff
@@ -200,10 +200,10 @@ describe( 'autosave', () => {
 		await page.evaluate( () => window.wp.data.dispatch( 'core/editor' ).__experimentalLocalAutosave() );
 		expect( await page.evaluate( () => window.sessionStorage.length ) ).toBe( 1 );
 
-		await page.reload();
+		await page.reload( { waitUntil: [ 'domcontentloaded', 'networkidle2' ] } );
+		await sleep( 2 );
```
Comment (gziolo) [inline packages/e2e-tests/specs/autosave.test.js]: Why does it need to wait 2 seconds after reload?

Comment (mcsf) [reply packages/e2e-tests/specs/autosave.test.js]: Was playing with reliability, expecting the failures to be related to a longer-than-expected data resolution for remote autosaves after the page has loaded.

With that in mind I just pushed 307f95a, which I'm more hopeful about. However, as I was waiting for Travis I just saw a test failing locally. (ノ°Д°）ノ︵ ┻━┻.

**Comment by gziolo** [comment]
Can you describe what the issue is and how you think it could be solved? Do you know why does this particular test fail? I can help debugging tomorrow.

**Comment by mcsf** [comment]
> Can you describe what the issue is and how you think it could be solved? Do you know why does this particular test fail? I can help debugging tomorrow.

Sure, and thanks for offering help.

From the logs, it seems that autosaving on the server occasionally just fails… because `editorSettings.autosave` should be available immediately upon page load, since it's provided by the server:

https://github.com/WordPress/WordPress/blob/099eb0578eecf79b7f25968011264c9e5d8014ba/wp-admin/edit-form-blocks.php#L302-L307

https://github.com/WordPress/WordPress/blob/099eb0578eecf79b7f25968011264c9e5d8014ba/wp-admin/edit-form-blocks.php#L385-L388

So this last change was a desperate attempt at retrying autosaving on the server.

File: packages/e2e-tests/specs/autosave.test.js [hunk 2]
```diff
@@ -196,14 +196,25 @@ describe( 'autosave', () => {
...
 		expect( await page.evaluate( () => window.sessionStorage.length ) ).toBe( 1 );
 
 		await page.reload();
 
+		// FIXME: Occasionally, upon reload, there is no server-provided
+		// autosave value available, despite our having previously explicitly
+		// autosaved. The reasons for this are still unknown. Since this is
+		// unrelated to *local* autosave, until we can understand them, we'll
+		// drop this test's expectations if we don't have an autosave object
+		// available.
```
Comment (youknowriad) [inline packages/e2e-tests/specs/autosave.test.js]: For me, we should just skip the test in this case instead of making it inconsistent.

Comment (youknowriad) [reply packages/e2e-tests/specs/autosave.test.js]: ideally, we find the real issue though :) @jasmussen had the issue while working on some PRs.

**Review by gziolo — APPROVED** [review]
It looks like a ton of work to improve the stability of this test. Major props for investing so much time to fully understand what the underlaying issue is.

I re-triggered Travis 3 times and it confirms that this fix is solid.

I don't want to block this PR with my own testing. Let's get this in and I will do my own experiments of the stable `master` branch. The reliability of the tests is the most important as it increases the confidence that Travis is useful :)

File: packages/e2e-tests/specs/autosave.test.js (hunk 2 above)
Comment (gziolo) [reply packages/e2e-tests/specs/autosave.test.js]: Yes, I will work on it today.

Comment (mcsf) [reply packages/e2e-tests/specs/autosave.test.js]: Thank you both!

Comment (jasmussen) [reply packages/e2e-tests/specs/autosave.test.js]: +1+1

**Comment by gziolo** [comment]
My further explorations performed in #17749 didn't help so far. We probably need to investigate whether autosave works properly in all cases. We might replicate what's used in one of the test where we intercept requests to gather more data where the issues might originate. See my related comment: https://github.com/WordPress/gutenberg/pull/17749#issuecomment-538018694.

**Comment by mcsf** [comment]
I agree, and thanks for digging!

File: packages/e2e-tests/specs/autosave.test.js (hunk 2 above)
Comment (aduth) [reply packages/e2e-tests/specs/autosave.test.js]: Was this worked on?

Still see the code:

https://github.com/WordPress/gutenberg/blob/5a5564b4d5cd6fe271ccb02e3f85f4141d7f21b5/packages/e2e-tests/specs/editor/various/autosave.test.js#L333-L346

Comment (mcsf) [reply packages/e2e-tests/specs/autosave.test.js]: It wasn't, and it still feels nebulous.

**Pull request merged**